open-craft/xblock-html
# fix: rendering tables [BB-6729]

### Description

This PR fixes issues related to tables generated by TinyMCE editor.

Tables are not rendered correctly in LMS - certain element, inline attributes and styles are not rendered, when JavaScript is not allowed for the XBlock. Enabling JavaScript fixes the issue, because it disables the sanitization of the rendered HTML. We still want the HTML to be sanitized, so we add the relevant elements, attributes and styles to the allow list for sanitization.

#### Screenshots
##### LMS
###### Before
![image](https://user-images.githubusercontent.com/30300520/191972812-ce28186c-3666-4edd-a5a8-417cb64ab1a4.png)
###### After
![image](https://user-images.githubusercontent.com/30300520/191973165-f594efdf-13ad-488c-b1a1-bfba3a9fa7a0.png)
##### Studio preview
###### Before
![image](https://user-images.githubusercontent.com/30300520/191972913-27720d2b-cadb-45fd-bbc0-f48812c3f115.png)
###### After
![image](https://user-images.githubusercontent.com/30300520/192473344-5b4f482f-5ff9-491c-8ba8-efc599d6555f.png)

### Testing instructions

1. Setup devstack (for development Maple was used, but Nutmeg or master should also work)
1. Clone `xblock-html` repo into `<devstack-root>/src/` and change the branch
1. Start `lms`, `studio`, and `learning` MFE:
   ```bash
    make lms-up studio-up frontend-app-learning-up
   ```
1. Install `xblock-html` into `lms`:
    ```bash
    make lms-shell lms-restart
    pip install -e /edx/src/xblock-html/ && exit
    ```
1. Install `xblock-html` into `studio`:
    ```bash
    make studio-shell studio-restart
    pip install -e /edx/src/xblock-html/ && exit
    ```
1. Login into studio as staff (e.g. `edx@example.com`)
1. Go to demo course advanced settings, and add `html5` and `excluded_html5` in `Advanced Module List`
1. Go to demo course outline, and create a new section, subsection and unit
1. In the new unit add `html5` xblock - Advanced > Text
1. Create a new table using the visual editor (or insert html provided below)
1. Save and publish
1. Check that the issues described in the description are fixed

The following HTML was generated using TinyMCE visual editor using all the settings which affect the tables styling:
```html
<table style="border-collapse: collapse; width: 100%; height: 34px; background-color: #c2e0f4; border-color: blue; border-style: solid; margin-left: auto; margin-right: auto;" cellspacing="1" cellpadding="1" border="10"><caption></caption>
<thead>
<tr style="height: 10px; background-color: #2dc26b; border-color: green; border-style: solid; text-align: center;">
<td style="width: 33.1779%; height: 10px;" scope="col"></td>
<td style="width: 33.1779%; height: 10px;" scope="col"></td>
<td style="width: 33.1779%; height: 10px;" scope="col"></td>
</tr>
</thead>
<tbody>
<tr style="height: 14px;">
<td style="width: 33.1779%; height: 14px;"></td>
<td style="width: 33.1779%; height: 14px; background-color: #3598db; text-align: center; vertical-align: middle; border: 2px solid #236fa1;"></td>
<td style="width: 33.1779%; height: 14px;"></td>
</tr>
</tbody>
<tfoot>
<tr style="height: 10px; background-color: #b96ad9; border-color: purple; border-style: solid; text-align: center;">
<td style="width: 33.1779%; height: 10px;"></td>
<td style="width: 33.1779%; height: 10px;"></td>
<td style="width: 33.1779%; height: 10px;"></td>
</tr>
</tfoot>
</table>
```

### Notes

I added a unit test for checking how html of table generated by TinyMCE would be rendered. I had to modify the html by sorting the inline attributes and styles in alphabetical order, so the it matches the sanitizer's output. The test has been useful during the development, but can be difficult to maintain, so can be removed from the PR.
